prov/gni/src/gnix_atomic.c

```diff
@@ -0,0 +1,606 @@
+{
+	if (dt >= FI_DATATYPE_LAST || op >= FI_ATOMIC_OP_LAST) {
+		return -FI_ENOENT;
+	}
+
+	switch(fr_type) {
+	case GNIX_FAB_RQ_AMO:
+		return __gnix_amo_cmds[op][dt] ?: -FI_ENOENT;
+	case GNIX_FAB_RQ_FAMO:
+		return __gnix_fetch_amo_cmds[op][dt] ?: -FI_ENOENT;
+	case GNIX_FAB_RQ_CAMO:
+		return __gnix_cmp_amo_cmds[op][dt] ?: -FI_ENOENT;
+	default:
+		break;
+	}
+
+	return -FI_ENOENT;
+}
+
+int _gnix_amo_post_req(void *data)
+{
+	struct gnix_fab_req *fab_req = (struct gnix_fab_req *)data;
+	struct gnix_fid_ep *ep = fab_req->gnix_ep;
+	struct gnix_nic *nic = ep->nic;
+	struct gnix_fid_mem_desc *loc_md;
+	struct gnix_tx_descriptor *txd;
+	gni_mem_handle_t mdh;
+	gni_return_t status;
+	int rc;
+	int inject_err = _gnix_req_inject_err(fab_req);
+
+	rc = _gnix_nic_tx_alloc(nic, &txd);
+	if (rc) {
+		GNIX_INFO(FI_LOG_EP_DATA, "_gnix_nic_tx_alloc() failed: %d\n",
+			 rc);
+		return -FI_EAGAIN;
+	}
+
+	txd->completer_fn = __gnix_amo_txd_complete;
+	txd->req = fab_req;
+
+	/* Mem handle CRC is not validated during FMA operations.  Skip this
+	 * costly calculation. */
+	_gnix_convert_key_to_mhdl_no_crc(
+			(gnix_mr_key_t *)&fab_req->amo.rem_mr_key,
+			&mdh);
+	loc_md = (struct gnix_fid_mem_desc *)fab_req->amo.loc_md;
+
+	txd->gni_desc.type = GNI_POST_AMO;
+	txd->gni_desc.cq_mode = GNI_CQMODE_GLOBAL_EVENT; /* check flags */
+	txd->gni_desc.dlvr_mode = GNI_DLVMODE_PERFORMANCE; /* check flags */
+	txd->gni_desc.local_addr = (uint64_t)fab_req->amo.loc_addr;
+	if (loc_md) {
+		txd->gni_desc.local_mem_hndl = loc_md->mem_hndl;
+	}
+	txd->gni_desc.remote_addr = (uint64_t)fab_req->amo.rem_addr;
+	txd->gni_desc.remote_mem_hndl = mdh;
+	txd->gni_desc.length = fab_req->amo.len;
+	txd->gni_desc.rdma_mode = 0; /* check flags */
+	txd->gni_desc.src_cq_hndl = nic->tx_cq; /* check flags */
+
+	txd->gni_desc.amo_cmd = _gnix_atomic_cmd(fab_req->amo.datatype,
+						 fab_req->amo.op,
+						 fab_req->type);
+	txd->gni_desc.first_operand = fab_req->amo.first_operand;
+	txd->gni_desc.second_operand = fab_req->amo.second_operand;
+
+	{
+		gni_mem_handle_t *tl_mdh = &txd->gni_desc.local_mem_hndl;
+		gni_mem_handle_t *tr_mdh = &txd->gni_desc.remote_mem_hndl;
+		GNIX_INFO(FI_LOG_EP_DATA, "la: %llx ra: %llx len: %d\n",
+			  txd->gni_desc.local_addr, txd->gni_desc.remote_addr,
+			  txd->gni_desc.length);
+		GNIX_INFO(FI_LOG_EP_DATA, "lmdh: %llx:%llx rmdh: %llx:%llx key: %llx\n",
+			  *(uint64_t *)tl_mdh, *(((uint64_t *)tl_mdh) + 1),
+			  *(uint64_t *)tr_mdh, *(((uint64_t *)tr_mdh) + 1),
+			  fab_req->amo.rem_mr_key);
+	}
+
+	fastlock_acquire(&nic->lock);
+
+	if (unlikely(inject_err)) {
+		_gnix_nic_txd_err_inject(nic, txd);
+		status = GNI_RC_SUCCESS;
+	} else {
+		status = GNI_PostFma(fab_req->vc->gni_ep, &txd->gni_desc);
+	}
+
+	fastlock_release(&nic->lock);
+
+	if (status != GNI_RC_SUCCESS) {
+		_gnix_nic_tx_free(nic, txd);
+		GNIX_INFO(FI_LOG_EP_DATA, "GNI_Post*() failed: %s\n",
+			  gni_err_str[status]);
+	}
+
+	return gnixu_to_fi_errno(status);
+}
+
+#if 0
```

Comment:
Why are you keeping this definition around? If this is meant for reference, it should be a comment.